vispy/visuals/filters/clipping_planes.py

```diff
@@ -0,0 +1,98 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) Vispy Development Team. All Rights Reserved.
+# Distributed under the (new) BSD License. See LICENSE.txt for more info.
+
+from functools import lru_cache
+
+import numpy as np
+
+from ..shaders import Function, Varying
+from .base_filter import Filter
+
+
+class PlanesClipper(Filter):
+    """Clips visual output based on arbitrary clipping planes."""
+
+    VERT_CODE = """
+    void clip() {
+        // Transform back to visual coordinates and clip based on that
+        $v_is_shown = $clip_with_planes($itransform(gl_Position).xyz);
+    }
+    """
+
+    FRAG_CODE = """
+    void clip() {
+        if ($v_is_shown < 0.)
+            discard;
+    }
+    """
+
+    def __init__(self, clipping_planes=None, coord_system='scene'):
+        tr = ['visual', 'scene', 'document', 'canvas', 'framebuffer', 'render']
+        if coord_system not in tr:
+            raise ValueError(f'Invalid coordinate system {coord_system}. Must be one of {tr}.')
+        self._coord_system = coord_system
+
+        super().__init__(
+            vcode=Function(self.VERT_CODE), vhook='post', vpos=1,
+            fcode=Function(self.FRAG_CODE), fhook='pre', fpos=1,
+        )
+
+        v_is_shown = Varying('v_is_shown', 'float')
+        self.vshader['v_is_shown'] = v_is_shown
+        self.fshader['v_is_shown'] = v_is_shown
+
+        self.clipping_planes = clipping_planes
+
+    @property
+    def coord_system(self):
+        # unsettable cause we can't update the transform after being attached
+        return self._coord_system
+
+    def _attach(self, visual):
+        super()._attach(visual)
+        self.vshader['itransform'] = visual.get_transform('render', self._coord_system)
+
+    @staticmethod
+    @lru_cache(maxsize=10)
+    def _build_clipping_planes_func(n_planes):
+        """Build the code snippet used to clip the volume based on self.clipping_planes."""
+        func_template = '''
+            float clip_planes(vec3 loc) {{
+                float is_shown = 3.4e38; // max float
```

Comment:
A bit of a nitpick, but using a bool for `is_shown` makes more sense semantically, and I don't think there's a technical reason not to do it, right?

Reply:
No, this intended: by having `is_shown` go from `+inf` to `-inf` and calculating it from the dot product, this allows us to correctly clip things like meshes in between vertices. This way, `is_shown` is correctly interpolated between two vertices on a fragment basis. Also, `varying` can only be a float or vec of floats.

Reply:
Ah, I think I understand it better now. So the comparison with the plane is done in the vertex shader, and the result is interpolated so the final decision can be made in the fragment shader.

Can we then change the name of `is_shown`? IIUC it represents the signed distance to the nearest plane, right?

Reply:
Yes! I'm fine with renaming; would you calling something like `signed_distance`?

Reply:
Maybe include the word "plane" or "clip", like `clip_distance` or something?

Reply:
`distance_from_clip`? Though it's actually the minimum value of all the signed distances...

Reply:
Acronym to `MSDFCP`? :P Ok, will go with that!